server.go

```diff
@@ -187,3 +224,196 @@ func copyWithHeader(r io.Reader, w io.Writer, header proto.Header) error {
 	}
 	return nil
 }
+
+func createSession(ctx context.Context, id string, command *Command, execer Execer, options *Options) (*Command, error) {
+	// If screen is not installed spawn the command normally.
+	_, err := exec.LookPath("screen")
+	if err != nil {
+		flog.Info("`screen` could not be found; session %s will not persist", id)
+		return command, nil
+	}
+
+	settings := []string{
+		// Tell screen not to handle motion for xterm* terminals which
+		// allows scrolling the terminal via the mouse wheel or scroll bar
+		// (by default screen uses it to cycle through the command history).
+		// There does not seem to be a way to make screen itself scroll on
+		// mouse wheel.  tmux can do it but then there is no scroll bar and
+		// it kicks you into copy mode where keys stop working until you
+		// exit copy mode which seems like it could be confusing.
+		"termcapinfo xterm* ti@:te@",
+		// Enable alternate screen emulation otherwise applications get
+		// rendered in the current window which wipes out visible output
+		// resulting in missing output when scrolling back with the mouse
+		// wheel (copy mode still works since that is screen itself
+		// scrolling).
+		"altscreen on",
+		// Remap the control key to C-s since C-a may be used in
+		// applications.  C-s cannot actually be used anyway since by
+		// default it will pause and C-q to resume will just kill the
+		// browser window.  We may not want people using the control key
+		// anyway since it will not be obvious they are in screen and doing
+		// things like switching windows makes mouse wheel scroll wonky due
+		// to the terminal doing the scrolling rather than screen itself
+		// (but again copy mode will work just fine).
+		"escape ^Ss",
+	}
+
+	dir := filepath.Join(os.TempDir(), "coder-screen")
+	config := filepath.Join(dir, "config")
+	screendir := filepath.Join(dir, "sockets")
+
+	err = os.MkdirAll(screendir, 0o700)
+	if err != nil {
+		return nil, xerrors.Errorf("unable to create %s for session %s: %w", screendir, id, err)
+	}
+
+	err = os.WriteFile(config, []byte(strings.Join(settings, "\n")), 0o644)
+	if err != nil {
+		return nil, xerrors.Errorf("unable to create %s for session %s: %w", config, id, err)
+	}
+
+	var sess *session
+	if rawSession, ok := sessions.Load(id); ok {
+		if sess, ok = rawSession.(*session); !ok {
+			return nil, xerrors.Errorf("found invalid type in session map for ID %s", id)
+		}
+	} else {
+		ctx, cancel := context.WithCancel(context.Background())
```

Comment:
you're shadowing `ctx` from the outer block and so it's really confusing to read what context is what --- pulling this into its own function would help make it much clearer.